lightning-persister/src/lib.rs

```diff
@@ -0,0 +1,416 @@
+		let persister = FilesystemPersister::new("test_rename_failure".to_string());
+		let test_writeable = TestWriteable{};
+		let txid_hex = "8984484a580b825b9972d7adb15050b3ab624ccd731946b3eeddb92f4e7ef6be";
+		let outp_idx = 0;
+		let test_txo = OutPoint {
+			txid: Txid::from_hex(txid_hex).unwrap(),
+			index: outp_idx,
+		};
+		// Create the channel data file and make it a directory.
+		let path = &persister.path_to_channel_data;
+		fs::create_dir_all(format!("{}/{}_{}", path, txid_hex, outp_idx)).unwrap();
+		match persister.write_channel_data(test_txo, &test_writeable) {
+			Err(e) => {
+				#[cfg(not(target_os = "windows"))]
+				assert_eq!(e.kind(), io::ErrorKind::Other);
+				#[cfg(target_os = "windows")]
+				assert_eq!(e.kind(), io::ErrorKind::PermissionDenied);
+			}
+			_ => panic!("Unexpected error message")
+		}
+	}
+
+	// Test failure to create the temporary file in the persistence process.
+	// We induce this failure by having the temp file already exist and be a
+	// directory.
+	#[test]
+	fn test_tmp_file_creation_failure() {
+		let persister = FilesystemPersister::new("test_tmp_file_creation_failure".to_string());
+		let test_writeable = TestWriteable{};
+		let txid_hex = "8984484a580b825b9972d7adb15050b3ab624ccd731946b3eeddb92f4e7ef6be";
+		let outp_idx = 0;
+		let test_txo = OutPoint {
+			txid: Txid::from_hex(txid_hex).unwrap(),
+			index: outp_idx,
+		};
+		// Create the tmp file and make it a directory.
+		let path = &persister.path_to_channel_data;
+		fs::create_dir_all(format!("{}/{}_{}.tmp", path, txid_hex, outp_idx)).unwrap();
+		match persister.write_channel_data(test_txo, &test_writeable) {
+			Err(e) => {
+				#[cfg(not(target_os = "windows"))]
+				assert_eq!(e.kind(), io::ErrorKind::Other);
+				#[cfg(target_os = "windows")]
+				assert_eq!(e.kind(), io::ErrorKind::PermissionDenied);
+			}
+			_ => panic!("Unexpected error message")
+		}
+	}
+
+	// Test that if the persister's path to channel data is read-only, writing a
+	// monitor to it results in the persister returning a PermanentFailure.
+	// Windows ignores the read-only flag for folders, so this test is Unix-only.
+	#[cfg(not(target_os = "windows"))]
+	#[test]
+	fn test_readonly_dir_perm_failure() {
+		let persister = FilesystemPersister::new("test_readonly_dir_perm_failure".to_string());
+		fs::create_dir_all(&persister.path_to_channel_data).unwrap();
+
+		// Set up a dummy channel and force close. This will produce a monitor
+		// that we can then use to test persistence.
+		let chanmon_cfgs = create_chanmon_cfgs(2);
+		let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+		let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+		let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+		let chan = create_announced_chan_between_nodes(&nodes, 0, 1, InitFeatures::known(), InitFeatures::known());
+		nodes[1].node.force_close_channel(&chan.2);
+		let mut added_monitors = nodes[1].chain_monitor.added_monitors.lock().unwrap();
```

Comment:
To avoid this setup, could you test the persister using `write_channel_data`like the earlier tests? Likewise in the next test.

Reply:
Hmm, it's nice to be able to test that it returns a `PermanentError` (since ultimately that's what we care about)